dashboard/modules/job/common.py

```diff
@@ -57,43 +58,93 @@ def __post_init__(self):
                 self.message = "Job failed."
 
 
-class JobStatusStorageClient:
+@dataclass
+class JobData:
+    status_info: JobStatusInfo
+    start_time: int
+    end_time: int = None
+    metadata: Optional[Dict[str, str]] = None
+    runtime_env: Optional[Dict[str, Any]] = None
+    namespace: Optional[str] = None
+
+
+class JobDataStorageClient:
     """
-    Handles formatting of status storage key given job id.
+    Interface to put and get job data from the Internal KV store.
     """
 
-    JOB_STATUS_KEY_PREFIX = "_ray_internal_job_status"
-    JOB_STATUS_KEY = f"{JOB_STATUS_KEY_PREFIX}_{{job_id}}"
+    JOB_DATA_KEY_PREFIX = "_ray_internal_job_data_"
+    JOB_DATA_KEY = f"{JOB_DATA_KEY_PREFIX}{{job_id}}"
 
     def __init__(self):
         assert _internal_kv_initialized()
 
-    def put_status(self, job_id: str, status: Union[JobStatus, JobStatusInfo]):
-        if isinstance(status, JobStatus):
-            status = JobStatusInfo(status=status)
-        elif not isinstance(status, JobStatusInfo):
-            assert False, "status must be JobStatus or JobStatusInfo."
-
+    def put_data(self, job_id: str, data: JobData):
         _internal_kv_put(
-            self.JOB_STATUS_KEY.format(job_id=job_id),
-            pickle.dumps(status),
+            self.JOB_DATA_KEY.format(job_id=job_id),
+            pickle.dumps(data),
             namespace=ray_constants.KV_NAMESPACE_JOB,
         )
 
-    def get_status(self, job_id: str) -> Optional[JobStatusInfo]:
-        pickled_status = _internal_kv_get(
-            self.JOB_STATUS_KEY.format(job_id=job_id),
+    def get_data(self, job_id: str) -> Optional[JobData]:
+        # raise ValueError
+        pickled_data = _internal_kv_get(
+            self.JOB_DATA_KEY.format(job_id=job_id),
             namespace=ray_constants.KV_NAMESPACE_JOB,
         )
-        if pickled_status is None:
+        if pickled_data is None:
+            return None
+        else:
+            return pickle.loads(pickled_data)
+
+    def put_status(self, job_id: str, status: Union[JobStatus, JobStatusInfo]):
+        """Puts or updates job status.  Sets end_time if status is terminal."""
+
+        if isinstance(status, JobStatus):
+            status_info = JobStatusInfo(status=status)
+        elif isinstance(status, JobStatusInfo):
+            status_info = status
+        else:
+            assert False, "status must be JobStatus or JobStatusInfo."
+
+        old_data = self.get_data(job_id)
+
+        if old_data is not None:
+            if (
+                status_info.status != old_data.status_info.status
+                and old_data.status_info.status.is_terminal()
+            ):
+                assert False, "Attempted to change job status from a terminal state."
+            new_data = replace(old_data, status_info=status_info)
```

Comment:
huh, i've never seen `replace` before. good to know...

Reply:
Yup, it also calls __post_init__ after the replacing is done.